runtime/src/main/cpp/Arrays.cpp

```diff
@@ -133,15 +133,30 @@ void Kotlin_Array_copyImpl(KConstRef thiz, KInt fromIndex,
     ThrowArrayIndexOutOfBoundsException();
   }
   mutabilityCheck(destination);
-  if (fromIndex >= toIndex) {
-    for (int index = 0; index < count; index++) {
-      UpdateHeapRef(ArrayAddressOfElementAt(destinationArray, toIndex + index),
-                      *ArrayAddressOfElementAt(array, fromIndex + index));
+  if (array == destinationArray && std::abs(fromIndex - toIndex) < count) {
+    // In case of coping inside same array number of decrements and increments of RC can be decreased.
+    if (fromIndex >= toIndex) {
+      // Relocated elements and release rewritten elements.
+      for (int index = 0; index < count; index++) {
+        UpdateHeapRefsInsideOneArray(array, destinationArray, index, fromIndex, toIndex, count);
+      }
+    } else {
+      // Relocated elements.
+      for (int index = count - 1; index >= 0; index--) {
+        UpdateHeapRefsInsideOneArray(array, destinationArray, index, fromIndex, toIndex, count);
+      }
     }
   } else {
-    for (int index = count - 1; index >= 0; index--) {
-      UpdateHeapRef(ArrayAddressOfElementAt(destinationArray, toIndex + index),
-                      *ArrayAddressOfElementAt(array, fromIndex + index));
+    if (fromIndex >= toIndex) {
```

Comment:
> Doesn't this if become redundant because we always copy elements between different arrays here?

Didn't get it, this if for order in loop.

Reply:
> CurrentMemoryModel != MemoryModel::kExperimental

Rewrote using this

Reply:
>Didn't get it, this if for order in loop.

As far as I understand, this order matters only if we copy elements inside the same array. Because otherwise we could rewrite elements that have not been copied yet. But for different arrays we can always use the same order because there is no danger to rewrite something. So this `if` looked redundant in the previous version of the patch.

But with the Alexander's suggestion we still use this logic in the new MM. So we do need this `if` in the current version of the patch.

Reply:
Ok, I got your idea, it still can be one array if we copy parts without intersection `abs(fromIndex - toIndex) > count`, but the order really doesn't matter. But yes, now it's part for new MM